src/parse/lexer.rs

```diff
@@ -0,0 +1,725 @@
+use crate::parse::error::{LexicalError, LexicalErrorType};
+use crate::parse::token::Tok;
+use std::char;
+
+// ideas:
```

Comment:
Is this comment still relevent?

Reply:
Good question. The lexer is so straight forward I'm not positive "look up 2 / 3 char tokens like keywords" would buy much but would pull the code size down considerably if it worked.

"track nesting per type and lex error on invalid nesting" Is about things like:

```
fn a() {
  ...
  )
}
```
and
```
fn a() {
  ...
  (
}
```

The lexer could keep track of where nestings were opened and closed rather easily on a stack. Every time a nesting is opened Its type and location are added to the stack, every time a nesting is closed it must complement the top of the stack. We could then emit a syntax error that pointed to the offending opening, closing, or both.

I did not fully implement this as I wasn't sure it was useful, though the lexer currently checks that there is at least _some_ nesting to close when you try, and makes sure that you don't end the module with any nestings open.

Reply:
I could see the error messages produced by stack being useful, but I'm not sure about the current bracket counting system as it seems a lot more limited.￼ Perhaps we could remove the bracket counting and add the stack approach in future if we deem that to be worth investing the time in?￼